src/common/tls.c

```diff
@@ -0,0 +1,191 @@
+/**
+ * (C) Copyright 2016-2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * It implements thread-local storage (TLS) for DAOS.
+ */
+#include <pthread.h>
+#include <daos/tls.h>
+
+/* The array remember all of registered module keys on one node. */
+struct daos_module_key *daos_module_keys[DAOS_MODULE_KEYS_NR] = { NULL };
+pthread_mutex_t daos_module_keys_lock = PTHREAD_MUTEX_INITIALIZER;
+
+void
+daos_register_key(struct daos_module_key *key)
+{
+	int i;
+
+	D_MUTEX_LOCK(&daos_module_keys_lock);
+	for (i = 0; i < DAOS_MODULE_KEYS_NR; i++) {
+		if (daos_module_keys[i] == NULL) {
+			daos_module_keys[i] = key;
+			key->dmk_index = i;
+			break;
+		}
+	}
+	D_MUTEX_UNLOCK(&daos_module_keys_lock);
+	D_ASSERT(i < DAOS_MODULE_KEYS_NR);
+}
+
+void
+daos_unregister_key(struct daos_module_key *key)
+{
+	if (key == NULL)
+		return;
+	D_ASSERT(key->dmk_index >= 0);
+	D_ASSERT(key->dmk_index < DAOS_MODULE_KEYS_NR);
+	D_MUTEX_LOCK(&daos_module_keys_lock);
+	daos_module_keys[key->dmk_index] = NULL;
+	D_MUTEX_UNLOCK(&daos_module_keys_lock);
+}
+
+/**
+ * Init thread context
+ *
+ * \param[in]dtls	Init the thread context to allocate the
```

Comment:
I don't think doxygen will like this, you need to adjust the whitespace.  Applies elsewhere as well.
```suggestion
 * \param[in] dtls	Init the thread context to allocate the
```

Reply:
sure